File: backend/app/schemas/team.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.profile import SkillLevel

TeamRole = Literal["head_coach", "assistant_coach", "captain", "player"]
TeamStatus = Literal["active", "without_coach"]
InviteKind = Literal["join", "head_coach"]
JoinRequestStatus = Literal["pending", "accepted", "rejected", "cancelled"]


class TeamIn(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str | None = Field(default=None, max_length=1000)
    sport: str = Field(min_length=1, max_length=50)
    age_category: str | None = Field(default=None, max_length=50)
    level: SkillLevel | None = None


class TeamOut(BaseModel):
    id: UUID
    name: str
    description: str | None
    sport: str
    age_category: str | None
    level: SkillLevel | None
    status: TeamStatus
    created_by: UUID
    logo_file_id: UUID | None = None
    created_at: datetime
    updated_at: datetime
    my_role: TeamRole | None = None
    members_count: int = 0


class TeamMemberOut(BaseModel):
    user_id: UUID
    telegram_id: int
    first_name: str
    last_name: str | None
    username: str | None
    photo_url: str | None
    role: TeamRole
    position: str | None
    joined_at: datetime


class MemberUpdateIn(BaseModel):
    role: Literal["captain", "assistant_coach", "player"] | None = None
    position: str | None = Field(default=None, max_length=50)


class TransferOwnershipIn(BaseModel):
    to_user_id: UUID
    confirmation_phrase: str


class InviteCreateIn(BaseModel):
    kind: InviteKind = "join"


class InviteOut(BaseModel):
    id: UUID
    team_id: UUID
    token: str
    kind: InviteKind
    link: str
    expires_at: datetime
    created_at: datetime


class JoinRequestOut(BaseModel):
    id: UUID
    team_id: UUID
    user_id: UUID
    first_name: str
    last_name: str | None
    username: str | None
    photo_url: str | None
    status: JoinRequestStatus
    created_at: datetime


class ApplyResultOut(BaseModel):
    status: Literal["pending", "joined"]
    team: TeamOut

File: frontend/src/types/team.ts
import type { SkillLevel } from "./profile";

export type TeamRole = "head_coach" | "assistant_coach" | "captain" | "player";
export type TeamStatus = "active" | "without_coach";
export type InviteKind = "join" | "head_coach";
export type JoinRequestStatus = "pending" | "accepted" | "rejected" | "cancelled";

export const TEAM_ROLE_LABELS: Record<TeamRole, string> = {
  head_coach: "Основной тренер",
  assistant_coach: "Помощник тренера",
  captain: "Капитан",
  player: "Игрок",
};

export interface Team {
  id: string;
  name: string;
  description: string | null;
  sport: string;
  ageCategory: string | null;
  level: SkillLevel | null;
  status: TeamStatus;
  createdBy: string;
  logoFileId: string | null;
  myRole: TeamRole | null;
  membersCount: number;
}

export interface TeamMember {
  userId: string;
  telegramId: number;
  firstName: string;
  lastName: string | null;
  username: string | null;
  photoUrl: string | null;
  role: TeamRole;
  position: string | null;
  joinedAt: string;
}

export interface Invite {
  id: string;
  teamId: string;
  token: string;
  kind: InviteKind;
  link: string;
  expiresAt: string;
  createdAt: string;
}

export interface JoinRequest {
  id: string;
  teamId: string;
  userId: string;
  firstName: string;
  lastName: string | null;
  username: string | null;
  photoUrl: string | null;
  status: JoinRequestStatus;
  createdAt: string;
}

export interface TeamDto {
  id: string;
  name: string;
  description: string | null;
  sport: string;
  age_category: string | null;
  level: SkillLevel | null;
  status: TeamStatus;
  created_by: string;
  logo_file_id: string | null;
  created_at: string;
  updated_at: string;
  my_role: TeamRole | null;
  members_count: number;
}

export interface TeamMemberDto {
  user_id: string;
  telegram_id: number;
  first_name: string;
  last_name: string | null;
  username: string | null;
  photo_url: string | null;
  role: TeamRole;
  position: string | null;
  joined_at: string;
}

export interface InviteDto {
  id: string;
  team_id: string;
  token: string;
  kind: InviteKind;
  link: string;
  expires_at: string;
  created_at: string;
}

export interface JoinRequestDto {
  id: string;
  team_id: string;
  user_id: string;
  first_name: string;
  last_name: string | null;
  username: string | null;
  photo_url: string | null;
  status: JoinRequestStatus;
  created_at: string;
}

export function mapTeamDto(dto: TeamDto): Team {
  return {
    id: dto.id,
    name: dto.name,
    description: dto.description,
    sport: dto.sport,
    ageCategory: dto.age_category,
    level: dto.level,
    status: dto.status,
    createdBy: dto.created_by,
    logoFileId: dto.logo_file_id,
    myRole: dto.my_role,
    membersCount: dto.members_count,
  };
}

export function mapTeamMemberDto(dto: TeamMemberDto): TeamMember {
  return {
    userId: dto.user_id,
    telegramId: dto.telegram_id,
    firstName: dto.first_name,
    lastName: dto.last_name,
    username: dto.username,
    photoUrl: dto.photo_url,
    role: dto.role,
    position: dto.position,
    joinedAt: dto.joined_at,
  };
}

export function mapInviteDto(dto: InviteDto): Invite {
  return {
    id: dto.id,
    teamId: dto.team_id,
    token: dto.token,
    kind: dto.kind,
    link: dto.link,
    expiresAt: dto.expires_at,
    createdAt: dto.created_at,
  };
}

export function mapJoinRequestDto(dto: JoinRequestDto): JoinRequest {
  return {
    id: dto.id,
    teamId: dto.team_id,
    userId: dto.user_id,
    firstName: dto.first_name,
    lastName: dto.last_name,
    username: dto.username,
    photoUrl: dto.photo_url,
    status: dto.status,
    createdAt: dto.created_at,
  };
}

export interface TeamInput {
  name: string;
  description: string | null;
  sport: string;
  age_category: string | null;
  level: SkillLevel | null;
}
